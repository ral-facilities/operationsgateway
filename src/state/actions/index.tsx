import {
  SettingsLoadedType,
  ConfigureURLsPayload,
  ConfigureURLsType,
} from './actions.types';
import { URLs, ActionType, ThunkResult } from '../state.types';
import { settings } from '../../settings';
import { Action } from 'redux';

export const settingsLoaded = (): Action => ({
  type: SettingsLoadedType,
});

export const loadUrls = (urls: URLs): ActionType<ConfigureURLsPayload> => ({
  type: ConfigureURLsType,
  payload: {
    urls,
  },
});

export const configureApp = (): ThunkResult<Promise<void>> => {
  return async (dispatch) => {
    const settingsResult = await settings;
    if (settingsResult) {
      dispatch(
        loadUrls({
          apiUrl: settingsResult['apiUrl'],
        })
      );

      dispatch(settingsLoaded());
    }
  };
};
