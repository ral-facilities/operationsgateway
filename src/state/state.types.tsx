import { ThunkAction } from 'redux-thunk';
import { AnyAction } from 'redux';

export interface URLs {
  apiUrl: string;
}

export interface OperationsGatewayState {
  urls: URLs;
}

export interface ActionType<T> {
  type: string;
  payload: T;
}

export type ThunkResult<R> = ThunkAction<
  R,
  OperationsGatewayState,
  null,
  AnyAction
>;
