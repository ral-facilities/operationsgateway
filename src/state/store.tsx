import {
  configureStore,
  combineReducers,
  PreloadedState,
} from '@reduxjs/toolkit';
import OperationsGatewayMiddleware, {
  listenToMessages,
} from './middleware/operationsgateway.middleware';
import configReducer from './slices/configSlice';
import tableReducer from './slices/tableSlice';
import searchReducer from './slices/searchSlice';
import plotReducer from './slices/plotSlice';

const rootReducer = combineReducers({
  config: configReducer,
  table: tableReducer,
  search: searchReducer,
  plots: plotReducer,
});

export function setupStore(preloadedState?: PreloadedState<RootState>) {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(OperationsGatewayMiddleware),
  });
}

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];

export const store = setupStore();

listenToMessages(store.dispatch);
