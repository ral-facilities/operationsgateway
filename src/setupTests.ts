// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import { Action } from 'redux';
import { Record } from './app.types';
import { initialState } from './state/reducers/operationsgateway.reducer';
import { OperationsGatewayState } from './state/state.types';

export let actions: Action[] = [];
export let resetActions = (): void => {
  actions = [];
};
export const getState = (): OperationsGatewayState => initialState;
export const dispatch = (action: Action): void | Promise<void> => {
  if (typeof action === 'function') {
    action(dispatch, getState);
    return Promise.resolve();
  } else {
    actions.push(action);
  }
};

export const flushPromises = (): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve));

export const testRecords: Record[] = [
  {
    id: '1',
    metadata: {
      dataVersion: '1',
      shotNum: 1,
      timestamp: '1',
      activeArea: '1',
      activeExperiment: '1',
    },
    channels: {
      test1: {
        metadata: {
          dataType: {
            units: 'km',
          },
        },
        data: 1,
      },
    },
  },
  {
    id: '2',
    metadata: {
      dataVersion: '1',
      shotNum: 2,
      timestamp: '2',
      activeArea: '2',
      activeExperiment: '2',
    },
    channels: {
      test2: {
        metadata: {
          dataType: {
            units: 'km',
          },
        },
        data: 2,
      },
    },
  },
  {
    id: '3',
    metadata: {
      dataVersion: '3',
      shotNum: 3,
      timestamp: '3',
      activeArea: '3',
      activeExperiment: '3',
    },
    channels: {
      test3: {
        metadata: {
          dataType: {
            units: 'km',
          },
        },
        data: 3,
      },
    },
  },
];
