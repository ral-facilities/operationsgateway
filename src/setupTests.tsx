// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
// need to mock <canvas> for plotting
import 'jest-canvas-mock';
import {
  Channel,
  FullChannelMetadata,
  ImageChannel,
  Record,
  RecordRow,
  ScalarChannel,
  WaveformChannel,
} from './app.types';
import { Action, PreloadedState, ThunkAction } from '@reduxjs/toolkit';
import { AppStore, RootState, setupStore } from './state/store';
import { initialState as initialConfigState } from './state/slices/configSlice';
import { initialState as initialTableState } from './state/slices/tableSlice';
import { initialState as initialSearchState } from './state/slices/searchSlice';
import { initialState as initialPlotState } from './state/slices/plotSlice';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { QueryClientProvider, QueryClient } from 'react-query';

// this is needed because of https://github.com/facebook/jest/issues/8987
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
let mockActualReact;
jest.doMock('react', () => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  if (!mockActualReact) {
    mockActualReact = jest.requireActual('react');
  }
  return mockActualReact;
});

export let actions: Action[] = [];
export const resetActions = (): void => {
  actions = [];
};
export const getInitialState = (): RootState => ({
  config: initialConfigState,
  table: initialTableState,
  search: initialSearchState,
  plot: initialPlotState,
});
export const dispatch = (
  action: Action | ThunkAction<void, RootState, unknown, Action<string>>
): void | Promise<void> => {
  if (typeof action === 'function') {
    action(dispatch, getInitialState, null);
    return Promise.resolve();
  } else {
    actions.push(action);
  }
};

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });

export const hooksWrapperWithProviders = (state = {}) => {
  const testQueryClient = createTestQueryClient();
  const store = setupStore(state);
  const wrapper = ({ children }) => (
    <Provider store={store}>
      <QueryClientProvider client={testQueryClient}>
        {children}
      </QueryClientProvider>
    </Provider>
  );
  return wrapper;
};

export function renderComponentWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  const testQueryClient = createTestQueryClient();
  function Wrapper({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element {
    return (
      <Provider store={store}>
        <QueryClientProvider client={testQueryClient}>
          {children}
        </QueryClientProvider>
      </Provider>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderComponentWithStore(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export const flushPromises = (): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve));

// MUI date pickers default to mobile versions during testing and so functions
// like .simulate('change') will not work, this workaround ensures desktop
// datepickers are used in tests instead
// https://github.com/mui/material-ui-pickers/issues/2073
export const applyDatePickerWorkaround = (): void => {
  // add window.matchMedia
  // this is necessary for the date picker to be rendered in desktop mode.
  // if this is not provided, the mobile mode is rendered, which might lead to unexpected behavior
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: (query: string) => ({
      media: query,
      // this is the media query that @material-ui/pickers uses to determine if a device is a desktop device
      matches: query === '(pointer: fine)',
      onchange: () => unknown,
      addEventListener: () => unknown,
      removeEventListener: () => unknown,
      addListener: () => unknown,
      removeListener: () => unknown,
      dispatchEvent: () => false,
    }),
  });
};

export const cleanupDatePickerWorkaround = (): void => {
  delete window.matchMedia;
};

export const testChannels: FullChannelMetadata[] = [
  {
    systemName: 'test_1',
    channel_dtype: 'scalar',
    userFriendlyName: 'Test 1',
    significantFigures: 4,
  },
  {
    systemName: 'test_2',
    channel_dtype: 'scalar',
    significantFigures: 2,
    scientificNotation: false,
  },
  {
    systemName: 'test_3',
    channel_dtype: 'scalar',
    significantFigures: 2,
    scientificNotation: true,
  },
];

export const generateRecord = (num: number): Record => {
  const numStr = `${num}`;

  let channel: Channel;

  if (num % 3 === 0) {
    channel = {
      metadata: {
        channel_dtype: 'scalar',
        units: 'km',
      },
      data:
        num < 10
          ? parseFloat(`${num}${num}${num}.${num}`)
          : parseFloat(
              numStr[0] + numStr[1] + numStr[1] + numStr[1] + '.' + numStr[1]
            ),
    } as ScalarChannel;
  } else if (num % 3 === 1) {
    channel = {
      metadata: {
        channel_dtype: 'image',
        horizontalPixels: num,
        horizontalPixelUnits: numStr,
        verticalPixels: num,
        verticalPixelUnits: numStr,
        cameraGain: num,
        exposureTime: num,
      },
      imagePath: numStr,
      thumbnail: numStr,
    } as ImageChannel;
  } else {
    channel = {
      metadata: {
        channel_dtype: 'waveform',
        xUnits: numStr,
        yUnits: numStr,
      },
      waveformId: numStr,
      thumbnail: numStr,
    } as WaveformChannel;
  }

  return {
    id: numStr,
    metadata: {
      dataVersion: numStr,
      shotnum: num,
      timestamp:
        num < 10 ? `2022-01-0${num} 00:00:00` : `2022-01-${num} 00:00:00`,
      activeArea: numStr,
      activeExperiment: numStr,
    },
    channels: {
      [`test_${num}`]: channel,
    },
  };
};

export const testRecords: Record[] = Array.from(Array(3), (_, i) =>
  generateRecord(i + 1)
);

export const generateRecordRow = (num: number) => {
  const record = generateRecord(num);

  const recordRow: RecordRow = {
    timestamp: record.metadata.timestamp,
    shotnum: record.metadata.shotnum,
    activeArea: record.metadata.activeArea,
    activeExperiment: record.metadata.activeExperiment,
  };

  const keys = Object.keys(record.channels);
  keys.forEach((key: string) => {
    const channel: Channel = record.channels[key];
    let channelData;
    const channelDataType = channel.metadata.channel_dtype;

    switch (channelDataType) {
      case 'scalar':
        channelData = (channel as ScalarChannel).data;
        break;
      case 'image':
        channelData = (channel as ImageChannel).thumbnail;
        channelData = (
          <img src={`data:image/jpeg;base64,${channelData}`} alt={key} />
        );
        break;
      case 'waveform':
        channelData = (channel as WaveformChannel).thumbnail;
        channelData = (
          <img src={`data:image/jpeg;base64,${channelData}`} alt={key} />
        );
    }

    recordRow[key] = channelData;
  });

  return recordRow;
};

export const testRecordRows = Array.from(Array(3), (_, i) =>
  generateRecordRow(i + 1)
);
