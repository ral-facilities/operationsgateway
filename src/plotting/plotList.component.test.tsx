import React from 'react';
import PlotList from './plotList.component';
import { PreloadedState } from '@reduxjs/toolkit';
import { renderComponentWithStore, getInitialState } from '../setupTests';
import { RootState } from '../state/store';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

describe('Plot Settings component', () => {
  let state: PreloadedState<RootState>;

  const createView = (initialState = state) => {
    return renderComponentWithStore(<PlotList />, {
      preloadedState: initialState,
    });
  };

  beforeEach(() => {
    state = getInitialState();
  });

  it('renders plot grid correctly with no plots', () => {
    const view = createView();

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('renders plot grid correctly with some plots', () => {
    state.plots = {
      'Plot 1': { open: true },
      'Plot 2': { open: false },
      'Plot 3': { open: true },
    };
    const view = createView();

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('creates a plot when user clicks create a plot button', async () => {
    const user = userEvent.setup();
    createView();

    await user.click(screen.getByRole('button', { name: 'Create a plot' }));

    expect(screen.getByText('Untitled 1')).toBeVisible();
  });

  it('deletes a plot when user clicks delete button', async () => {
    state.plots = {
      'Plot 1': { open: false },
    };
    const user = userEvent.setup();
    createView();

    expect(screen.getByText('Plot 1')).toBeVisible();

    await user.click(screen.getByRole('button', { name: 'Delete' }));

    expect(screen.queryByText('Plot 1')).not.toBeInTheDocument();
  });

  it('opens a plot when user clicks open button', async () => {
    state.plots = {
      'Plot 1': { open: false },
    };
    const user = userEvent.setup();
    const { store } = createView();

    expect(screen.getByText('Plot 1')).toBeVisible();

    await user.click(screen.getByRole('button', { name: 'Edit' }));

    expect(store.getState().plots['Plot 1']?.open).toBe(true);
  });
});
