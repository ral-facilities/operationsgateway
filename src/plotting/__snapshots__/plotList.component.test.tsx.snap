// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plot Settings component renders plot grid correctly with no plots 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-column css-16ybxmp-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Create a plot
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-4 css-13twzmd-MuiGrid-root"
    />
  </div>
</DocumentFragment>
`;

exports[`Plot Settings component renders plot grid correctly with some plots 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-column css-16ybxmp-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Create a plot
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-4 css-13twzmd-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            Plot 1
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall css-1knaqv7-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Edit
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall css-1knaqv7-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Delete
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            Plot 2
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall css-1knaqv7-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Edit
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall css-1knaqv7-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Delete
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            Plot 3
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall css-1knaqv7-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Edit
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall css-1knaqv7-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Delete
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
